Streamable HTTP: optional GET/SSE stream rejected with 404 kills the whole MCP connection

## Describe the bug

After a Streamable HTTP MCP server accepts the mandatory POST-based handshake, Qwen Code's client also probes an optional standalone `GET <mcp-url>` with `Accept: text/event-stream` (the server-push notification stream, optional per the MCP spec). If the server rejects that optional probe with `HTTP 404` instead of `400`, the connection is torn down entirely and the server is reported as "failed to start" — even though the mandatory POST-based communication works perfectly and the server is fully spec-compliant.

This reproduces against **third-party MCP servers unrelated to any one implementation**, so it isn't specific to a particular backend:

- `https://mcp.context7.com/mcp` — POST `initialize` succeeds (`200`), no auth needed.
- `https://api.githubcopilot.com/mcp/` — POST `initialize` returns `401` (needs a token) same as a healthy server would.

Both return `404` for `GET <url>` with `Accept: text/event-stream`, and both end up in the "MCP server(s) failed to start" warning.

## Root cause

`STREAMABLE_HTTP_GET_SSE_FALLBACK_STATUSES` in `packages/core/src/tools/mcp-client.ts` only tolerates `400` for this optional-stream rejection (added in #4521 to support Spring AI's specific behavior). Servers that instead return the arguably more conventional `404` (no route registered for `GET`) aren't covered, so the rejection propagates as a fatal `SseError` and the whole MCP client connection is aborted instead of continuing with POST-only communication.

## To Reproduce

1. Configure any MCP server over `httpUrl` (Streamable HTTP) whose `GET <url>` (no session, `Accept: text/event-stream`) returns `404` — e.g. point at `https://mcp.context7.com/mcp` directly:
   ```json
   { "mcpServers": { "context7": { "httpUrl": "https://mcp.context7.com/mcp" } } }
   ```
2. Run `qwen -p "say hi" --debug` (or start an interactive session).
3. Observe:
   ```
   Warning: MCP server(s) failed to start: context7. Continuing with built-in tools and any servers that did connect.
   ```
4. With `QWEN_CODE_DEBUG=1`, the debug log (`~/.qwen/debug/<session>.txt`) shows:
   ```
   [ERROR] [MCP] MCP ERROR (context7): SSE error: Non-200 status code (404)
   [ERROR] [MCP] Error during discovery for server 'context7': SSE error: Non-200 status code (404)
   ```

Verified independently with `curl`:
```
$ curl -s -o /dev/null -w "%{http_code}\n" -X POST https://mcp.context7.com/mcp \
    -H "Content-Type: application/json" -H "Accept: application/json, text/event-stream" \
    -d '{"jsonrpc":"2.0","id":1,"method":"initialize","params":{"protocolVersion":"2025-06-18","capabilities":{},"clientInfo":{"name":"test","version":"0.0.1"}}}'
200
$ curl -s -o /dev/null -w "%{http_code}\n" -X GET https://mcp.context7.com/mcp -H "Accept: text/event-stream"
405
```
(context7 itself returns `405`; a handful of other Streamable-HTTP-only servers we tested return `404` for the same GET — either way, neither is in the current `Set([400])`.)

## Expected behavior

A server rejecting the *optional* GET/SSE stream with any conventional "not supported here" status (`400`, `404`, or the transport's own native `405` sentinel) should not tear down the MCP connection — POST-based tool calls should keep working, same as the existing `400` (Spring AI) handling from #4521.

## Environment

- Qwen Code version: `0.21.8` (latest on npm at time of filing)
- OS: Windows 11 (also reproduces on the bundled CLI's Node/undici fetch path, not shell/proxy specific)
- Install method: `npm install -g @qwen-code/qwen-code`

## Additional context

This also affects self-hosted Streamable HTTP MCP servers built with the official TypeScript SDK's stateless `StreamableHTTPServerTransport` pattern (a plain Express `app.get('/mcp', ...)` handler returning `404`/`405` for the unsupported GET) — i.e. any team standing up their own Streamable HTTP-only MCP server following the SDK's documented stateless example is affected, not just servers with unusual behavior.

I have a minimal patch ready (extends the existing `Set` to also cover `404`, deliberately leaving `405` untouched since it's already handled natively elsewhere and rewriting it would break existing test coverage for that path) and I'm opening a PR shortly for review.

--- a/packages/core/src/tools/mcp-client.test.ts
+++ b/packages/core/src/tools/mcp-client.test.ts
@@ -2949,6 +2949,115 @@ lOTTGqPpwFUbw2EMOOpFYuIyzGMIpUNMBjE2gvJiqFQ=
         expect(await response.text()).toBe('');
       });
 
+      it('treats 404 from optional GET SSE stream as unsupported', async () => {
+        const fetchFn = vi.fn<typeof fetch>().mockResolvedValue(
+          // Streamable HTTP servers with no GET route at all reject the optional
+          // standalone GET/SSE notification stream with 404 — e.g. the official
+          // MCP SDK's documented stateless StreamableHTTPServerTransport pattern
+          // behind Express, where 404 is Express's own default fallthrough for
+          // the unhandled GET (#8784). (Note: context7 returns a raw 405, which
+          // the SDK tolerates natively — the earlier claim that it returned 404
+          // was retracted in the issue as a reporter-side misconfiguration.)
+          new Response('not found', { status: 404 }),
+        );
+        const fetchWithFallback = createStreamableHttpCompatibilityFetch(
+          'no-get-route',
+          fetchFn,
+        );
+
+        const response = await fetchWithFallback('http://test-server/mcp', {
+          method: 'GET',
+          headers: { Accept: 'text/event-stream' },
+        });
+
+        expect(fetchFn).toHaveBeenCalledTimes(1);
+        expect(response.status).toBe(405);
+        expect(response.statusText).toBe('Method Not Allowed');
+        expect(await response.text()).toBe('');
+      });
+
+      it('does not rewrite non-SSE GET 404 responses', async () => {
+        const fetchFn = vi
+          .fn<typeof fetch>()
+          .mockResolvedValue(new Response('not found', { status: 404 }));
+        const fetchWithFallback = createStreamableHttpCompatibilityFetch(
+          'plain-get-404',
+          fetchFn,
+        );
+
+        const response = await fetchWithFallback('http://test-server/mcp', {
+          method: 'GET',
+          headers: { Accept: 'application/json' },
+        });
+
+        expect(response.status).toBe(404);
+        expect(await response.text()).toBe('not found');
+      });
+
+      it('does not rewrite POST 404 responses', async () => {
+        const fetchFn = vi
+          .fn<typeof fetch>()
+          .mockResolvedValue(new Response('not found', { status: 404 }));
+        const fetchWithFallback = createStreamableHttpCompatibilityFetch(
+          'post-404',
+          fetchFn,
+        );
+
+        // The SDK's real POST requests set this exact Accept header
+        // (streamableHttp.ts's transport always sends
+        // 'application/json, text/event-stream'), so the method check is
+        // the only thing standing between a genuine tool-call 404 and being
+        // silently rewritten into a synthetic 405 — the Accept header alone
+        // does not disambiguate POST from the optional GET/SSE probe.
+        const response = await fetchWithFallback('http://test-server/mcp', {
+          method: 'POST',
+          headers: {
+            'Content-Type': 'application/json',
+            Accept: 'application/json, text/event-stream',
+          },
+        });
+
+        expect(response.status).toBe(404);
+        expect(await response.text()).toBe('not found');
+      });
+
+      it('bounds the fallback body excerpt read when the 404 body stalls', async () => {
+        // A server that answers the optional GET/SSE probe with 404 headers and
+        // then never sends a body chunk must not park the diagnostics read
+        // forever: the MCP dispatcher runs with `headersTimeout: 0,
+        // bodyTimeout: 0` and nothing above this wrapper bounds the body.
+        vi.useFakeTimers();
+        try {
+          const fetchFn = vi.fn<typeof fetch>().mockResolvedValue(
+            new Response(
+              new ReadableStream({
+                // Headers are already delivered; the body never yields.
+                pull: () => new Promise<void>(() => {}),
+              }),
+              { status: 404 },
+            ),
+          );
+          const fetchWithFallback = createStreamableHttpCompatibilityFetch(
+            'stalled-404-body',
+            fetchFn,
+          );
+
+          const pending = fetchWithFallback('http://test-server/mcp', {
+            method: 'GET',
+            headers: { Accept: 'text/event-stream' },
+          });
+          await vi.advanceTimersByTimeAsync(2_000);
+          const response = await pending;
+
+          expect(response.status).toBe(405);
+          expect(mockDebugLogger.warn).toHaveBeenCalledWith(
+            expect.not.stringContaining('Response body:'),
+          );
+        } finally {
+          vi.useRealTimers();
+        }
+      });
+
       it('omits response body diagnostics when the fallback body is empty', async () => {
         const fetchFn = vi
           .fn<typeof fetch>()

--- a/packages/core/src/tools/mcp-client.ts
+++ b/packages/core/src/tools/mcp-client.ts
@@ -128,8 +128,15 @@ function bindInvocationContextPolicy(
   }
 }
 
-const STREAMABLE_HTTP_GET_SSE_FALLBACK_STATUSES = new Set([400]);
+const STREAMABLE_HTTP_GET_SSE_FALLBACK_STATUSES = new Set([400, 404]);
 const STREAMABLE_HTTP_GET_SSE_ERROR_BODY_LIMIT = 512;
+// The MCP undici dispatcher runs with headersTimeout: 0, bodyTimeout: 0
+// (see getOrCreateMcpDispatcher in runtimeFetchOptions.ts), and every caller
+// of the optional GET/SSE probe is fire-and-forget, so nothing above this
+// wrapper will ever time out a body that sends headers but never completes.
+// This excerpt is best-effort diagnostics only — bound it so a stalled body
+// can't leak a reader/connection for the life of the process.
+const STREAMABLE_HTTP_GET_SSE_EXCERPT_TIMEOUT_MS = 2_000;
 
 export function getMcpOAuthDialogInstruction(
   action: 'authenticate' | 're-authenticate',
@@ -177,11 +184,11 @@ async function readResponseBodyExcerpt(
     return undefined;
   }
 
-  const decoder = new TextDecoder();
-  let body = '';
-  let bytesRead = 0;
-  let truncated = false;
-  try {
+  const readLoop = async (): Promise<string | undefined> => {
+    const decoder = new TextDecoder();
+    let body = '';
+    let bytesRead = 0;
+    let truncated = false;
     while (bytesRead < STREAMABLE_HTTP_GET_SSE_ERROR_BODY_LIMIT) {
       const { done, value } = await reader.read();
       if (done) {
@@ -225,7 +232,23 @@ async function readResponseBodyExcerpt(
       excerpt.length > STREAMABLE_HTTP_GET_SSE_ERROR_BODY_LIMIT
       ? `${excerpt.slice(0, STREAMABLE_HTTP_GET_SSE_ERROR_BODY_LIMIT)}...`
       : excerpt;
+  };
+
+  try {
+    return await runWithTimeout(
+      readLoop(),
+      STREAMABLE_HTTP_GET_SSE_EXCERPT_TIMEOUT_MS,
+      'Streamable HTTP GET SSE fallback body excerpt',
+    );
   } catch {
+    // Timed out (or the read failed): drop the diagnostics and cancel this
+    // clone's tee branch so the abandoned read settles instead of dangling.
+    // Cancelling one branch alone does not run the underlying source's
+    // cancel algorithm — it's the caller's response.body?.cancel() below
+    // that cancels the sibling branch and actually releases the connection.
+    reader.cancel().catch(() => {
+      // Best-effort cleanup; the caller still returns the 405 sentinel.
+    });
     return undefined;
   }
 }
@@ -250,8 +273,14 @@ function isStreamableHttpGetSseRequest(init?: RequestInit): boolean {
 
 /**
  * Wraps fetch to preserve OAuth challenges before the SDK discards response
- * metadata and to normalize Spring AI-style 400 responses to the SDK's
- * unsupported sentinel for the optional Streamable HTTP GET SSE request.
+ * metadata and to normalize conventional "GET not supported" rejections of
+ * the optional Streamable HTTP GET SSE request to the SDK's unsupported
+ * sentinel: Spring AI rejects it with 400 (#4521), and servers with no GET
+ * route at all — e.g. the official SDK's stateless
+ * `StreamableHTTPServerTransport` behind Express, whose default fallthrough
+ * answers 404 — reject it with 404 (#8784). A raw 405 needs no rewriting
+ * (the SDK tolerates it natively) and 401 must stay untouched (the OAuth
+ * challenge detection above depends on observing it).
  *
  * SDK coupling: `StreamableHTTPClientTransport._startOrAuthSse()` treats a
  * 405 response as "GET SSE unsupported" and continues in POST-only mode.

